dashboard/test/models/contact_rollups_pardot_memory_test.rb

```diff
@@ -32,52 +32,167 @@ class ContactRollupsPardotMemoryTest < ActiveSupport::TestCase
+      {
+        email: 'beta',
+        pardot_id: 2, pardot_id_updated_at: base_time - 2.days,
+        data_synced_at: base_time - 1.day, data_synced: {db_Opt_In: 'Yes'}
+      },
+      {
+        email: 'gamma',
+        pardot_id: 3, pardot_id_updated_at: base_time + 2.days,
+        data_synced_at: base_time + 1.day, data_synced: {db_Opt_In: 'Yes'}
+      },
+      # dummy records
+      {email: 'delta', pardot_id: nil},
+      {email: 'epsilon', pardot_id: 4},
+    ]
+    pardot_memory_records.each {|record| create :contact_rollups_pardot_memory, record}
+
+    # 3 cases that require updating contacts
+    processed_contact_records = [
+      # Case 1: data_synced_at is null
+      {email: 'alpha', data: {opt_in: false}, data_updated_at: base_time},
+      # Case 2: data_updated_at > data_synced_at
+      {email: 'beta', data: {opt_in: false}, data_updated_at: base_time},
+      # Case 3: pardot_id_updated_at > data_synced_at
+      {email: 'gamma', data: {opt_in: true}, data_updated_at: base_time},
+      # dummy records
+      {email: 'delta'},
+      {email: 'zeta'},
+    ]
+    processed_contact_records.each {|record| create :contact_rollups_processed, record}
+
+    # Execute SQL query
+    results = ActiveRecord::Base.connection.
+      exec_query(ContactRollupsPardotMemory.query_updated_contacts).map do |record|
+      record.slice('email', 'pardot_id_changed')
     end
+
+    expected_results = [
+      {'email' => 'alpha', 'pardot_id_changed' => nil},
+      {'email' => 'beta', 'pardot_id_changed' => 0},
+      {'email' => 'gamma', 'pardot_id_changed' => 1}
+    ]
+    assert_equal expected_results, results
   end
 
-  test 'save_sync_results' do
+  test 'update_pardot_prospects' do
+    email = 'test@domain.com'
+    last_sync_time = Time.now.utc - 7.days
+    create :contact_rollups_pardot_memory, email: email, data_synced: {db_Opt_In: 'No'}, data_synced_at: last_sync_time
+    create :contact_rollups_processed, email: email, data: {'opt_in' => true}
+
+    PardotV2.expects(:submit_batch_request).once.returns([])
+
+    ContactRollupsPardotMemory.update_pardot_prospects
+
+    record = ContactRollupsPardotMemory.find_by(email: email)
+    assert_equal({'db_Opt_In' => 'Yes'}, record&.data_synced)
+    assert last_sync_time < record&.data_synced_at
+  end
+
+  test 'save_sync_results new prospect' do
     assert_equal 0, ContactRollupsPardotMemory.count
 
-    submissions = [
-      {email: 'invalid_email', id: nil, db_Opt_in: 'No'},
-      {email: 'valid@domain.com', id: nil, db_Opt_in: 'Yes'},
+    submission = {email: 'valid@domain.com', db_Opt_In: 'Yes'}
+    submitted_time = Time.now.utc
+
+    ContactRollupsPardotMemory.save_sync_results [submission], [], submitted_time
+
+    record = ContactRollupsPardotMemory.find_by(
+      email: submission[:email],
+      data_synced_at: submitted_time
+    )
+    expected_data_synced = submission.except(:email, :id).deep_stringify_keys
```

Comment:
No ID in the submission any more?

Reply:
`data_synced` column contains only contact metadata/attributes. There are separate `email` and `pardot_id` columns. Will also remove the safe navigator in the assertion below.